packages/grafana-ui/src/components/Table/hooks.ts

```diff
@@ -36,41 +36,47 @@ export function useFixScrollbarContainer(
 }
 
 /**
-  react-table caches the height of cells so we need to reset them when expanding/collapsing rows
-   We need to take the minimum of the current expanded indexes and the previous expandedIndexes array to account
-  for collapsed rows, since they disappear from expandedIndexes but still keep their expanded height
+  react-table caches the height of cells, so we need to reset them when expanding/collapsing rows.
+  We use lastExpandedOrCollapsedIndex since collapsed rows disappear from expandedIndexes but still keep their expanded
+  height.
  */
 export function useResetVariableListSizeCache(
   extendedState: GrafanaTableState,
   listRef: React.RefObject<VariableSizeList>,
-  data: DataFrame
+  data: DataFrame,
+  uniqueField?: Field
 ) {
+  // Make sure we trigger the reset when keys change in any way
+  const expandedRowsRepr = JSON.stringify(Object.keys(extendedState.expanded));
+
   useEffect(() => {
-    if (extendedState.lastExpandedIndex !== undefined) {
-      // Gets the expanded row with the lowest index. Needed to reset all expanded row heights from that index on
-      let resetIndex = extendedState.lastExpandedIndex;
-      const expandedIndexes = Object.keys(extendedState.expanded);
-      if (expandedIndexes.length > 0) {
-        const lowestExpandedIndex = parseInt(expandedIndexes[0], 10);
-        if (!isNaN(lowestExpandedIndex)) {
-          resetIndex = Math.min(resetIndex, lowestExpandedIndex);
-        }
+    // By default, reset all rows
+    let resetIndex = 0;
+
+    // If we have unique field, extendedState.expanded keys are not row indexes but IDs so instead of trying to search
+    // for correct index we just reset the whole table.
+    if (!uniqueField) {
+      // If we don't have we reset from the last changed index.
+      if (Number.isFinite(extendedState.lastExpandedOrCollapsedIndex)) {
```

Comment:
This is simplified from previous version. Previously it reset always the minimal index whether it was `lastExpandedOrCollapsedIndex` or from `state.expanded`. This right now only expands from last changed index in `lastExpandedOrCollapsedIndex` down.